File: truezip-kernel/src/main/java/de/schlichtherle/truezip/kernel/BasicArchiveController.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truezip.kernel;

import static de.truezip.kernel.FsAccessOption.APPEND;
import static de.truezip.kernel.FsAccessOption.CREATE_PARENTS;
import static de.truezip.kernel.FsAccessOptions.NONE;
import de.truezip.kernel.*;
import de.truezip.kernel.cio.Entry.Access;
import static de.truezip.kernel.cio.Entry.Access.READ;
import static de.truezip.kernel.cio.Entry.Access.WRITE;
import de.truezip.kernel.cio.Entry.Type;
import static de.truezip.kernel.cio.Entry.Type.DIRECTORY;
import static de.truezip.kernel.cio.Entry.Type.FILE;
import de.truezip.kernel.cio.*;
import de.truezip.kernel.io.InputException;
import de.truezip.kernel.io.Streams;
import de.truezip.kernel.util.BitField;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.FileAlreadyExistsException;
import java.nio.file.FileSystemException;
import java.nio.file.NoSuchFileException;
import java.util.Map;
import java.util.Objects;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.annotation.CheckForNull;
import javax.annotation.concurrent.NotThreadSafe;

/**
 * An abstract base class for any archive file system controller which
 * provide all the essential services required for accessing a prospective
 * archive file.
 * This base class encapsulates all the code which is not depending on a
 * particular archive update strategy and the corresponding state of this
 * file system controller.
 * <p>
 * Each instance of this class manages an archive file - the <i>target file</i>
 * - in order to allow random access to it as if it were a regular directory in
 * its parent file system.
 * <p>
 * Note that in general all of the methods in this class are reentrant on
 * exceptions.
 * This is important because client applications may repeatedly call them.
 * Of course, depending on the calling context, some or all of the archive
 * file's data may be lost in this case.
 * 
 * @param  <E> the type of the archive entries.
 * @author Christian Schlichtherle
 */
@NotThreadSafe
abstract class BasicArchiveController<E extends FsArchiveEntry>
extends LockModelController {

    private static final Logger logger = Logger.getLogger(
            BasicArchiveController.class.getName(),
            BasicArchiveController.class.getName());

    /**
     * Constructs a new basic archive controller.
     *
     * @param model the non-{@code null} archive model.
     */
    BasicArchiveController(final LockModel model) {
        super(model);
        if (null == model.getParent())
            throw new IllegalArgumentException();
    }

    /** Equivalent to {@link #autoMount(boolean) autoMount(false)}. */
    final ArchiveFileSystem<E> autoMount(BitField<FsAccessOption> options)
    throws IOException {
        return autoMount(options, false);
    }

    /**
     * Returns the (virtual) archive file system mounted from the target
     * archive file.
     *
     * @param  options the options for accessing the file system entry.
     * @param  autoCreate If this is {@code true} and the archive file does not
     *         exist, then a new archive file system with only a virtual root
     *         directory is created with its last modification time set to the
     *         system's current time.
     * @return An archive file system.
     * @throws IOException on any I/O error.
     */
    abstract ArchiveFileSystem<E> autoMount(
            BitField<FsAccessOption> options,
            boolean autoCreate)
    throws IOException;

    @Override
    public final boolean isReadOnly() throws IOException {
        return autoMount(NONE).isReadOnly();
    }

    @Override
    public final FsEntry stat(
            BitField<FsAccessOption> options,
            FsEntryName name)
    throws IOException {
        return autoMount(options).stat(options, name);
    }

    @Override
    public final void checkAccess(
            BitField<FsAccessOption> options,
            FsEntryName name,
            BitField<Access> types)
    throws IOException {
        autoMount(options).checkAccess(options, name, types);
    }

    @Override
    public final void setReadOnly(FsEntryName name) throws IOException {
        autoMount(NONE).setReadOnly(name);
    }

    @Override
    public final boolean setTime(
            BitField<FsAccessOption> options,
            FsEntryName name,
            Map<Access, Long> times)
    throws IOException {
        checkSync(options, name, null);
        return autoMount(options).setTime(options, name, times);
    }

    @Override
    public final boolean setTime(
            BitField<FsAccessOption> options,
            FsEntryName name,
            BitField<Access> types, long value)
    throws IOException {
        checkSync(options, name, null);
        return autoMount(options).setTime(options, name, types, value);
    }

    @Override
    public final InputSocket<?> input(
            final BitField<FsAccessOption> options,
            final FsEntryName name) {
        Objects.requireNonNull(options);
        Objects.requireNonNull(name);

        @NotThreadSafe
        class Input extends DelegatingInputSocket<E> {
            @Override
            public E localTarget() throws IOException {
                peerTarget(); // may sync() if in same target archive file!
                checkSync(options, name, READ);
                final FsCovariantEntry<E> ce = autoMount(options).stat(options, name);
                if (null == ce)
                    throw new NoSuchFileException(name.toString());
                return ce.getEntry();
            }

            @Override
            protected InputSocket<E> socket() throws IOException {
                final FsArchiveEntry ae = localTarget();
                final Type type = ae.getType();
                if (FILE != type)
                    throw new FileSystemException(name.toString(), null,
                            "Expected a FILE entry, but is a " + type + " entry!");
                return input(ae.getName());
            }
        }
        return new Input();
    }

    abstract InputSocket<E> input(String name);

    @Override
    public final OutputSocket<?> output(
            final BitField<FsAccessOption> options,
            final FsEntryName name,
            final @CheckForNull Entry template) {
        Objects.requireNonNull(options);
        Objects.requireNonNull(name);

        @NotThreadSafe
        class Output extends AbstractOutputSocket<FsArchiveEntry> {
            @Override
            public FsArchiveEntry localTarget() throws IOException {
                final E ae = mknod().get().getEntry();
                if (options.get(APPEND)) {
                    // A proxy entry must get returned here in order to inhibit
                    // a peer target to recognize the type of this entry and
                    // switch to Raw Data Copy (RDC) mode.
                    // This would not work when APPENDing.
                    return new ProxyEntry(ae);
                }
                return ae;
            }

            @Override
            @edu.umd.cs.findbugs.annotations.SuppressWarnings("RCN_REDUNDANT_NULLCHECK_OF_NULL_VALUE") // false positive
            public OutputStream stream() throws IOException {
                final ArchiveFileSystemOperation<E> op = mknod();
                final E ae = op.get().getEntry();
                final InputStream in = append();
                Throwable ex = null;
                try {
                    final OutputSocket<? extends E> os = output(options, ae);
                    if (null == in) // do NOT bind when appending!
                        os.bind(this);
                    final OutputStream out = os.stream();
                    try {
                        op.commit();
                        if (null != in)
                            Streams.cat(in, out);
                    } catch (final Throwable ex2) {
                        try {
                            out.close();
                        } catch (final Throwable ex3) {
                            ex2.addSuppressed(ex3);
                        }
                        throw ex2;
                    }
                    return out;
                } catch (final Throwable ex2) {
                    ex = ex2;
                    throw ex2;
                } finally {
                    if (null != in) {
                        try {
                            in.close();
                        } catch (final IOException ex2) {
                            final IOException ex3 = new InputException(ex2);
                            if (null == ex)
                                throw ex3;
                            ex.addSuppressed(ex3);
                        } catch (final Throwable ex2) {
                            if (null == ex)
                                throw ex2;
                            ex.addSuppressed(ex2);
                        }
                    }
                }
            }

            ArchiveFileSystemOperation<E> mknod() throws IOException {
                checkSync(options, name, WRITE);
                // Start creating or overwriting the archive entry.
                // This will fail if the entry already exists as a directory.
                return autoMount(options, !name.isRoot() && options.get(CREATE_PARENTS))
                        .mknod(options, name, FILE, template);
            }

            @CheckForNull InputStream append() {
                if (options.get(APPEND)) {
                    try {
                        return input(options, name).stream();
                    } catch (IOException ignored) {
                        // When appending, there is no need for the entry to be
                        // readable or even exist, so this can get safely ignored.
                    }
                }
                return null;
            }
        }
        return new Output();
    }

    private static final class ProxyEntry
    extends DecoratingEntry<FsArchiveEntry>
    implements FsArchiveEntry {
        ProxyEntry(FsArchiveEntry entry) {
            super(entry);
        }

        @Override
        public Type getType() {
            return entry.getType();
        }

        @Override
        public boolean setSize(Size type, long value) {
            return entry.setSize(type, value);
        }

        @Override
        public boolean setTime(Access type, long value) {
            return entry.setTime(type, value);
        }

        @Override
        public Boolean isPermitted(Access type, Entity entity) {
            return entry.isPermitted(type, entity);
        }

        @Override
        public boolean setPermitted(Access type, Entity entity, Boolean value) {
            return entry.setPermitted(type, entity, value);
        }
    } // ProxyEntry

    abstract OutputSocket<E> output(BitField<FsAccessOption> options, E entry);

    @Override
    public final void mknod(
            final BitField<FsAccessOption> options,
            final FsEntryName name,
            final Type type,
            final Entry template)
    throws IOException {
        if (name.isRoot()) { // TODO: Is this case differentiation still required?
            try {
                autoMount(options); // detect false positives!
            } catch (final FalsePositiveArchiveException ex) {
                if (DIRECTORY != type)
                    throw ex;
                autoMount(options, true);
                return;
            }
            throw new FileAlreadyExistsException(name.toString(), null,
                    "Cannot replace a directory entry!");
        } else {
            checkSync(options, name, null);
            autoMount(options, options.get(CREATE_PARENTS))
                    .mknod(options, name, type, template)
                    .commit();
        }
    }

    @Override
    public void unlink(
            final BitField<FsAccessOption> options,
            final FsEntryName name)
    throws IOException {
        checkSync(options, name, null);
        final ArchiveFileSystem<E> fs = autoMount(options);
        fs.unlink(options, name);
        if (name.isRoot()) {
            // Check for any archive entries with absolute entry names.
            final int size = fs.size() - 1; // mind the ROOT entry
            if (0 != size)
                logger.log(Level.WARNING, "unlink.absolute",
                        new Object[] { getMountPoint(), size });
        }
    }

    /**
     * Checks if the intended access to the named archive entry in the virtual
     * file system is possible without performing a
     * {@link FsController#sync(BitField, ExceptionHandler) sync} operation in
     * advance.
     *
     * @param  options the options for accessing the file system entry.
     * @param  name the name of the file system entry.
     * @param  intention the intended I/O operation on the archive entry.
     *         If {@code null}, then only an update to the archive entry meta
     *         data (i.e. a pure virtual file system operation with no I/O)
     *         is intended.
     * @throws NeedsSyncException If a sync operation is required before the
     *         intended access could succeed.
     */
    abstract void checkSync(
            BitField<FsAccessOption> options,
            FsEntryName name,
            @CheckForNull Access intention)
    throws NeedsSyncException;
}

File: truevfs-kernel/src/main/scala/de/schlichtherle/truevfs/kernel/se/BasicArchiveController.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truevfs.kernel.se

import de.schlichtherle.truevfs.kernel._
import net.truevfs.kernel._
import net.truevfs.kernel.FsAccessOption._
import net.truevfs.kernel.FsAccessOptions._
import net.truevfs.kernel.cio._
import net.truevfs.kernel.cio.Entry._;
import net.truevfs.kernel.cio.Entry.Access._;
import net.truevfs.kernel.cio.Entry.Type._;
import net.truevfs.kernel.io._
import net.truevfs.kernel.util._
import java.io._
import java.nio.channels._
import java.nio.file._
import java.util.logging._

/**
 * An abstract base class for any archive file system controller which
 * provide all the essential services required for accessing a prospective
 * archive file.
 * This base class encapsulates all the code which is not depending on a
 * particular archive update strategy and the corresponding state of this
 * file system controller.
 * <p>
 * Each instance of this class manages an archive file - the <i>target file</i>
 * - in order to allow random access to it as if it were a regular directory in
 * its parent file system.
 * <p>
 * Note that in general all of the methods in this class are reentrant on
 * exceptions.
 * This is important because client applications may repeatedly call them.
 * Of course, depending on the calling context, some or all of the archive
 * file's data may be lost in this case.
 * 
 * @param  <E> the type of the archive entries.
 * @author Christian Schlichtherle
 */
private abstract class BasicArchiveController[E <: FsArchiveEntry](m: LockModel)
extends FsAbstractController[LockModel](m) {
  import BasicArchiveController._

  require(null ne m.getParent)

  override def stat(options: AccessOptions, name: FsEntryName): FsEntry =
    autoMount(options).stat(options, name).orNull

  override def checkAccess(options: AccessOptions, name: FsEntryName, types: BitField[Access]) =
    autoMount(options).checkAccess(options, name, types)

  override def setReadOnly(name: FsEntryName) =
    autoMount(NONE).setReadOnly(name)

  override def setTime(options: AccessOptions, name: FsEntryName, times: java.util.Map[Access, java.lang.Long]) = {
    checkSync(options, name, None)
    autoMount(options).setTime(options, name, times)
  }

  override def setTime(options: AccessOptions, name: FsEntryName, types: BitField[Access], value: Long) = {
    checkSync(options, name, None)
    autoMount(options).setTime(options, name, types, value)
  }

  override def input(options: AccessOptions, name: FsEntryName) = {
    require(null ne options)
    require(null ne name)

    final class Input extends DelegatingInputSocket[E] {
      override def socket() = {
        val ae = localTarget()
        val tµpe = ae.getType()
        if (FILE ne tµpe)
          throw new FileSystemException(name.toString, null,
                                        "Expected a FILE entry, but is a " + tµpe + " entry!");
        input(ae.getName)
      }

      override def localTarget() = {
        peerTarget() // may sync() if in same target archive file!
        checkSync(options, name, Some(READ))
        autoMount(options).stat(options, name) match {
          case Some(ce) => ce.getEntry
          case _ => throw new NoSuchFileException(name.toString)
        }
      }
    }
    new Input
  }: AnyInputSocket

  def input(name: String): InputSocket[E]

  override def output(options: AccessOptions, name: FsEntryName, template: Entry) = {
    require(null ne options)
    require(null ne name)

    final class Output extends AbstractOutputSocket[FsArchiveEntry] {
      override def localTarget() = {
        val ae = mknod().head.getEntry
        if (options.get(APPEND)) {
          // A proxy entry must get returned here in order to inhibit
          // a peer target to recognize the type of this entry and
          // switch to Raw Data Copy (RDC) mode.
          // This would not work when APPENDing.
          new ProxyEntry(ae)
        } else {
          ae
        }
      }

      override def stream() = {
        val tx = mknod()
        val ae = tx.head.getEntry
        val in = append()
        var ex: Option[Throwable] = None
        try {
          val os = output(options, ae)
          in match {
            case None => os.bind(this)
            case _ => // do NOT bind when appending!
          }
          val out = os.stream()
          try {
            tx.commit()
            in.foreach(Streams.cat(_, out))
          } catch {
            case ex2: Throwable =>
              try {
                out.close()
              } catch {
                case ex3: Throwable =>
                  ex2.addSuppressed(ex3)
              }
              throw ex2
          }
          out
        } catch {
          case ex2: Throwable =>
            ex = Some(ex2)
            throw ex2
        } finally {
          in.foreach { in =>
            try {
              in.close()
            } catch {
              case ex2: IOException =>
                val ex3 = new InputException(ex2)
                ex match {
                  case Some(ex) => ex.addSuppressed(ex3)
                  case _ => throw ex3
                }
              case ex2: Throwable =>
                ex match {
                  case Some(ex) => ex.addSuppressed(ex2)
                  case _ => throw ex2
                }
            }
          }
        }
      }

      def mknod() = {
        checkSync(options, name, Some(WRITE))
        // Start creating or overwriting the archive entry.
        // This will fail if the entry already exists as a directory.
        autoMount(options, !name.isRoot && options.get(CREATE_PARENTS))
        .mknod(options, name, FILE, Option(template))
      }

      def append(): Option[InputStream] = {
        if (options.get(APPEND)) {
          try {
            return Some(input(options, name).stream())
          } catch {
            // When appending, there is no need for the entry to be
            // readable or even exist, so this can get safely ignored.
            case _: IOException =>
          }
        }
        None
      }
    }
    new Output
  }: AnyOutputSocket

  def output(options: AccessOptions, entry: E): OutputSocket[E]

  override def mknod(options: AccessOptions, name: FsEntryName, tµpe: Type, template: Entry): Unit = {
    if (name.isRoot) { // TODO: Is this case differentiation still required?
      try {
        autoMount(options) // detect false positives!
      } catch {
        case ex: FalsePositiveArchiveException =>
          if (DIRECTORY ne tµpe)
            throw ex
          autoMount(options, true);
          return
      }
      throw new FileAlreadyExistsException(name.toString, null,
                                           "Cannot replace a directory entry!");
    } else {
      checkSync(options, name, None)
      autoMount(options, options.get(CREATE_PARENTS))
      .mknod(options, name, tµpe, Option(template))
      .commit()
    }
  }

  override def unlink(options: AccessOptions, name: FsEntryName) = {
    checkSync(options, name, None)
    val fs = autoMount(options)
    fs.unlink(options, name)
    if (name.isRoot) {
      // Check for any archive entries with absolute entry names.
      val size = fs.size - 1 // mind the ROOT entry
      if (0 != size)
          logger.log(Level.WARNING, "unlink.absolute",
                     Array[AnyRef](getMountPoint, size.asInstanceOf[AnyRef]));
    }
  }

  /**
   * Checks if the intended access to the named archive entry in the virtual
   * file system is possible without performing a
   * {@link FsController#sync(BitField, ExceptionHandler) sync} operation in
   * advance.
   *
   * @param  options the options for accessing the file system entry.
   * @param  name the name of the file system entry.
   * @param  intention the intended I/O operation on the archive entry.
   *         If {@code None}, then only an update to the archive entry meta
   *         data (i.e. a pure virtual file system operation with no I/O)
   *         is intended.
   * @throws NeedsSyncException If a sync operation is required before the
   *         intended access could succeed.
   */
  def checkSync(options: AccessOptions, name: FsEntryName, intention: Option[Access])

  /**
   * Returns the (virtual) archive file system mounted from the target
   * archive file.
   *
   * @param  options the options for accessing the file system entry.
   * @param  autoCreate If this is {@code true} and the archive file does not
   *         exist, then a new archive file system with only a virtual root
   *         directory is created with its last modification time set to the
   *         system's current time.
   * @return An archive file system.
   * @throws IOException on any I/O error.
   */
  def autoMount(options: AccessOptions, autoCreate: Boolean = false): ArchiveFileSystem[E]
}

private object BasicArchiveController {
  private val logger = Logger.getLogger(
    classOf[BasicArchiveController[_]].getName,
    classOf[BasicArchiveController[_]].getName)

  private final class ProxyEntry(entry: FsArchiveEntry)
  extends DecoratingEntry[FsArchiveEntry](entry) with FsArchiveEntry {
    override def getType: Type = entry.getType

    override def setSize(tµpe: Size, value: Long) = entry.setSize(tµpe, value)

    override def setTime(tµpe: Access, value: Long) = entry.setTime(tµpe, value)

    override def isPermitted(tµpe: Access, entity: Entity) =
      entry.isPermitted(tµpe, entity)

    override def setPermitted(tµpe: Access, entity: Entity, value: java.lang.Boolean) =
      entry.setPermitted(tµpe, entity, value)
  }
}
